system/lib/compiler-rt/emscripten_setjmp.c

```diff
@@ -102,4 +103,28 @@ void __wasm_longjmp(void *env, int val) {
   __builtin_wasm_throw(C_LONGJMP, &__wasm_longjmp_args);
 }
 
+// jmp_buf should have large enough size and alignment to contain
+// this structure.
+struct jmp_buf_impl {
+  void* func_invocation_id;
+  uint32_t label;
+};
+
+void __wasm_setjmp(void* env, uint32_t label, void* func_invocation_id) {
+  struct jmp_buf_impl* jb = env;
+  assert(label != 0);                 // ABI contract
+  assert(func_invocation_id != NULL); // sanity check
+  jb->func_invocation_id = func_invocation_id;
+  jb->label = label;
+}
+
+uint32_t __wasm_setjmp_test(void* env, void* func_invocation_id) {
+  struct jmp_buf_impl* jb = env;
+  assert(jb->label != 0);             // ABI contract
+  assert(func_invocation_id != NULL); // sanity check
+  if (jb->func_invocation_id == func_invocation_id) {
+    return jb->label;
+  }
+  return 0;
+}
```

Comment:
Does the corresponding LLVM change not also change the ABI for emscripten exceptions too?  i.e. not just when __USING_WASM_SJLJ__?  (i.e. should all this new code live outside the `#if` block?

Reply:
These `__wasm_setjmp` and `__wasm_setjmp_test` can be used both for emscripten and Wasm SjLj. Actually, using different functions for the two of them will be messy and not allow us to delete the existing `saveSetjmp` and `testSetjmp`. This is currently within `#ifdef __USING_WASM_SJLJ__`. Only `__wasm_longjmp` is exclusive to Wasm SjLj.

Reply:
> Does the corresponding LLVM change not also change the ABI for emscripten exceptions too? i.e. not just when **USING_WASM_SJLJ**? (i.e. should all this new code live outside the `#if` block?

currently, no.

Reply:
> +1. These `__wasm_setjmp` and `__wasm_setjmp_test` can be used both for emscripten and Wasm SjLj. Actually, using different functions for the two of them will be messy and not allow us to delete the existing `saveSetjmp` and `testSetjmp`. This is currently within `#ifdef __USING_WASM_SJLJ__`. Only `__wasm_longjmp` is exclusive to Wasm SjLj.

maybe. i haven't looked at the emscrpiten sjlj closely. do you think it's enough to apply the same change mechanically?